doc/npc.md

```diff
@@ -0,0 +1,559 @@
+# NotifyPropertyChanged aspect for Metalama
+This document provides implementation overview for PostSharp's `NotifyPropertyChanged` aspect, lists important features, analyses performance,
+"nice-to-have" features that were not implemented by the original implementation but requested frequently by users and finally discusses 
+possible implementation strategies.
+
+# Current Implementation
+`NotifyPropertyChanged` aspect automatically implements `INotifyPropertyChanged` and `INotifyPropertyChanging` interfaces with PropertyChanged
+and `PropertyChanging` events.
+
+In contrary to a naive implementation, it also attempts to solve the following:
+1) Raise PropertyChanged event only when the object is in a consistent state (i.e. delay the event until the control leaves the object's method).
+2) Do not raise PropertyChanged multiple times when changed multiple times within one method (implied by 1).
+3) Do not raise PropertyChanged when the value of the property has not changed (optional).
+4) Allow subscribers to events to be GC'ed, i.e. weak event semantics (optional).
+
+## Aspect interface
+The aspect implements an "internal" interface `INotifyChildPropertyChanged`, which is the main mean of communication between instances of the aspect.
+
+It allows to propagate changes to properties of child objects upstream using `ChildPropertyChanged` event, which carries a "property path", a sequence of property names which describe the source of the change.
+
+```
+class A
+{
+    public B Property { get; set; }
+}
+
+class B
+{
+    public int AnotherProperty { get;set }
+}
+```
+
+In the above example, object `x` of type `A` can raise `ChildPropertyChanged` with `Property.AnotherProperty` argument, which would mean that `x.Property.AnotherProperty` changed.
+
+```
+class C
+{
+    public A A { get; set; }
+
+    public int YetAnotherProperty => this.A.Property.AnotherProperty;
+}
+```
+
+An object `c` of class `C` above subscribes to `ChildPropertyChanged` of `c.A` property. When it receives a notification for `Property.AnotherProperty` from `c.A`, it means that `c.A.Property.AnotherProperty` has changed and PropertyChanged should be raised for `c.YetAnotherProperty`.
+
+## Compile-time analysis
+The aspect automatically analyses the user code of the target class. The analysis is performed on PostSharp's code model, which builds expressions on top of the IL code. 
+
+However, due to limitations of PostSharp, the analysis is performed only on the original code, all other aspects need to integrate with `NotifyPropertyChanged` aspect using internal interfaces and helpers.

```

Comment:
Automatic dependency analysis (an important improvement from a naive implementation) is limited to the code in the input assembly, i.e. the user code. When implementing other aspects that need to integrate with the NPC aspect, the integration must be explicitly implemented.

We have the same situation at the moment in Metalama (no way to analyze result of template expansion that may contain additional dependencies). It would be useful to have a good programmatic way of integrating with the aspect anyway.